frame/im-online/src/lib.rs

```diff
@@ -315,13 +315,32 @@ decl_module! {
 
 		fn deposit_event() = default;
 
-		#[weight = MINIMUM_WEIGHT]
+		/// `keys_len` must be the length for the vec stored in storage `Keys`.
+		///
+		/// # <weight>
+		/// - Complexity: `O(K + E)` where K is length of `Keys` and E is length of
+		///   `Heartbeat.network_state.external_address`
+		///
+		///   - `O(K)`: decoding of length `K`
+		///   - `O(E)`: decoding/encoding of length `E`
+		/// - DbReads: pallet_session `Validators`, pallet_session `CurrentIndex`, `Keys`,
+		///   `ReceivedHeartbeats`
+		/// - DbWrites: `ReceivedHeartbeats`
+		/// # </weight>
+		// NOTE: the weight include cost of validate_unsigned as it is part of the cost to import
+		// block with such an extrinsic.
+		#[weight = 310_000_000
+			+ 750_000 * *_keys_len as Weight
+			+ 1_200_000 * heartbeat.network_state.external_addresses.len() as Weight
+			+ T::DbWeight::get().reads_writes(4, 1)]
 		fn heartbeat(
```

Comment:
This weight could be rewrite using `Keys::decode_len` instead of `_keys_len` but that would allow someone to send a signed extrinsic which will make this storage query: `Keys::decode_len` while not having funds to pay for it.

Thus I follow the rule: no storage read in weights.